test/shared/src/test/scala/zio/test/mock/RandomSpec.scala

```diff
@@ -4,7 +4,7 @@ import scala.concurrent.Future
 import scala.util.{ Random => SRandom }
 
 import zio.{ Chunk, DefaultRuntime, UIO }
-import zio.test.mock.MockRandom.{ DefaultData, Mock }
+import zio.test.mock.MockRandom.{ DefaultData, Mock => RMock }
```

Comment:
We may want to find a new name for the `Mock` objects in `MockRandom` and friends to avoid namespace conflicts and confusion.

Reply:
Maybe we should rename `Mock` to `Proxy`? This way, `MockRandom.Mock`, for example, can stay the same?

Reply:
I think we should keep the name `Mock` for this and rename current impl to `Fake` or `Stub` following the semantics defined [here](https://martinfowler.com/articles/mocksArentStubs.html).

Reply:
TLDR;

> - **Dummy** objects are passed around but never actually used. Usually they are just used to fill parameter lists.
> - **Fake** objects actually have working implementations, but usually take some shortcut which makes them not suitable for production (an in memory database is a good example).
> - **Stubs** provide canned answers to calls made during the test, usually not responding at all to anything outside what's programmed in for the test.
> - **Spies** are stubs that also record some information based on how they were called. One form of this might be an email service that records how many messages it was sent.
> - **Mocks** are what we are talking about here: objects pre-programmed with expectations which form a specification of the calls they are expected to receive.

Reply:
I am open to this but I am not sold on this level of precision because most people will not have read Fowler on the topic.

@adamgfraser Thoughts?

Reply:
I like the idea of having different names for the existing `Mock` implementations versus this new functionality because I do think they facilitate significantly different types of testing. For example, a "mock" implementation of `Clock` provided by this PR is going to behave very differently from the existing `MockClock` implementation. So it would be nice to have different names and possibly different packages within `test` to help us make this distinction and communicate it clearly to users in our documentation.

That said, I have a couple of concerns with the specifics of this framework:
- I think `Fake` and `Stub` names have negative connotations we want to avoid. I'm obviously somewhat biased here but I think the existing `Mock` implementations, especially `MockClock` and `MockRandom`, provide some nice functionality and would like to be able to talk about them. I think it is a lot easier to say "we have this awesome MockClock implementation" than we have an awesome "FakeClock" or "StubClock". Maybe it will actually make more sense to people who have read Fowler but a lot of people won't have.
- I think the full framework is too much complexity to educate users on. I think we can say there are two categories of things, the existing Mock implementations and this new mocking functionality, whatever we call them. I think the full five categories make things too complex, especially since some of our existing implementations cross the boundaries between the categories.
- We should at least be aware that if we change the names of the existing Mock implementations there is some cost to users and the education effort. There is no good time to change a name and sooner is probably better than later but we are just starting to educate people about using these things so changing the name right now is not the greatest.

My suggestion was to rename the existing `MockClock` and so on to `TestClock`. I realize this is a little overly general since everything we're doing has to do with testing, but I think it is a clear and non-perjorative and frees up the "mock" namespace for this new functionality.

Reply:
I prefer `Test*` too.